_posts/2024-07-28-Uncovering-Code-Coverage.md

```diff
@@ -0,0 +1,229 @@
+author: dave-bunten
+tags:
+  - code-coverage
+  - test-coverage
+  - software-quality
+  - testing
+  - software-sustainability
+---
+
+# Uncovering Code Coverage: Ensuring Software Reliability with Comprehensive Testing
+
+{% include blog-post-intro.html %}
+
+## Introduction
+
+{% include figure.html image="images/file-holding-umbrella-for-coverage.png" width="200px" caption="Code coverage gives our code protection from the unknown." %}
+
+<!-- excerpt start -->
+
+Test coverage is a crucial aspect of software development that helps ensure your code is reliable and bug-free.
+By measuring how much of your code is covered by tests, you can identify untested areas and improve overall quality.
+In this post, we'll dive into what test coverage is, why it matters, and explore some tools for measuring it.
+Let's get started!
+
+<!-- excerpt end -->
+
+## What is Code Coverage?
+
+{% include figure.html image="images/simple-example-of-coverage-for-functions.png" width="600px" caption="Code which is tested is considered 'covered'." %}
+
+> "With tests, we can change the behavior of our code quickly and verifiably. Without them, we really don't know if our code is getting better or worse." ([Michael Feathers, Working Effectively with Legacy Code](https://www.goodreads.com/quotes/718460-code-without-tests-is-bad-code-it-doesn-t-matter-how))
+
+Code coverage, also known as test coverage, refers to the percentage of your code that is executed when your tests run ([Wikipedia: Code coverage](https://en.wikipedia.org/wiki/Code_coverage)).
+Software developers use code coverage as a way to understand areas which may need tests.
+By understanding these metrics, you can get a clearer picture of your code's reliability.
+
+## Types of Coverage
+
+```python
+# example line
+print("A line which could be tested for coverage.")
+
+# example statement (potentially multi-line)
+statement = [
+    1,
+    2,
+    3,
+]
+
+# example function
+def example():
+    return "An example function."
+
+# example branching
+if 1 in statement:
+    return True
+else:
+    return False
+```
+
+Line coverage, statement coverage, function coverage, and branch coverage common metrics used to measure how thoroughly code is tested, but they each focus on different aspects.
+
+- **Line coverage** tracks the execution of individual lines of code, indicating whether each line has been run during testing.
+- **Statement coverage** is similar but considers whether each statement (a single unit of execution) within the code has been executed, regardless of the number of lines.
+- **Function coverage** checks whether each function or method in the code has been called during the tests, ensuring that all parts of the code's interface are exercised.
+- **Branch coverage**, on the other hand, goes deeper by verifying whether every possible path through control structures (like if-else statements) has been followed, ensuring that all potential outcomes and decision points are tested. 
```

Comment:
Might be overthinking this, but "on the other hand, goes deeper by" suggested to me that this was the  most thorough of all the options. If I understand correctly, line and statement coverage would always ensure all branches are covered.

Maybe reword to "Branch coverage goes deeper as compared to function coverage"... or just skip the preamble since the other bullets don't have it and just say "Branch coverage verifies whether".

Reply:
Thanks, I'll reword this for clarity. Generally I think each method goes deep or shallow in certain areas. Branch coverage does go a bit deeper with concerns to branches than statement-level coverage; if-else blocks are generally treated as one compound statement, meaning that when checking statement coverage any branch triggered could show the full if-else block as covered (despite there being branches which perhaps are untested).